Porting zbox sdk and set minimum function

Rizary opened the pull request:
This is initial work on porting all sdk and zcncore function to wasm. In this PR, I have add:
1. Create new folder `wasm` and move all `_sdkver` into `wasm`
2. Update README on how to build wasm and test the function
3. move function that belong to wasm/sdk.go from wasm/proxy.go
4. Adding go test on `wasm/sdk.go` and `wasm/wallet.go` for non-http calls function

When this branched is merged to `jssdk`, we can merge `jssdk` to `master`

moldis commented on bls/bls.go:
```diff
@@ -97,7 +98,7 @@ func ToBytes(E *BN254.ECP) []byte {
 func ToBytes2(E *BN254.ECP2) []byte {
 	const MFS = BN254.MFS
-	const G1S = 2*MFS + 1 /* Group 1 Size */
+	// const G1S = 2*MFS + 1 /* Group 1 Size */
```
Please, remove all comments from your code. Except comments for function headers

Rizary replied:
since this is golangci-lint problem, I'll fix in the next commit.

moldis commented on bls/bls.go:
```diff
@@ -442,7 +443,7 @@ func (sk *SecretKey) GetPublicKey() *PublicKey {
 	const MGS = BN254.MGS
 	const MFS = BN254.MFS
-	const G1S = 2*MFS + 1 /* Group 1 Size */
+	// const G1S = 2*MFS + 1 /* Group 1 Size */
```
Btw why it was removed from here?

cnlangzi replied:
We can't remove variables exported. Because it might be used by 3rd package. golang-lint can't detect it properly.

moldis commented on core/logger/logger.go:
```diff
@@ -15,14 +15,15 @@ const (
-const cCyan = "\u001b[36m"
-const cWhite = "\u001b[37m"
+
+// const cGreen = "\u001b[32m"
```
+1 to comments

moldis commented on core/logger/logger.go:
```diff
@@ -85,25 +86,25 @@ func (l *Logger) SetLogFile(logFile io.Writer, verbose bool) {
 func (l *Logger) Info(v ...interface{}) {
 	if l.lvl >= INFO {
-		l.logInfo.Output(2, fmt.Sprint(v...))
+		_ = l.logInfo.Output(2, fmt.Sprint(v...))
```
If you not using value, you don't need  `_ = `

moldis commented on core/transaction/entity.go:
```diff
@@ -237,10 +238,10 @@ func VerifyTransaction(txnHash string, sharders []string) (*Transaction, error)
-				//Logger.Error("Error unmarshalling to get transaction response", err.Error())
-			}
+			_ = json.Unmarshal(objmap["txn"], txn)
+			// if err != nil {
```
Why this check was removed?

moldis commented on core/util/merkle_tree.go:
```diff
@@ -116,7 +116,7 @@ func (mt *MerkleTree) VerifyPath(hash Hashable, path *MTPath) bool {
 
 /*GetPathByIndex - get the path of a leaf node at index i */
 func (mt *MerkleTree) GetPathByIndex(idx int) *MTPath {
-	path := make([]string, mt.levels-1, mt.levels-1)
```
Why it was removed?

moldis commented on core/zcncrypto/bls0chain_test.go:
```diff
@@ -115,20 +115,21 @@ func TestSerialization(t *testing.T) {
-	testSetByCSPRNGCase(t, []byte{178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178, 178}, "1e2520a9b2b2b2abc9e17cb2b2b2b2912e2eb2b2b2b2b2 ...
-	testSetByCSPRNGCase(t, []byte{5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5}, "0505050505050505050505050505050505050505050505050505050505050505")
-}
+// func TestSetByCSPRNG(t *testing.T) {
```
Why this test commented?

Rizary replied:
I got some linter error said that it isn't used. I should put `//nolint` in there. Will fix in the next commit.

cnlangzi replied:
@Rizary  it is test. it should be used by `go test`.

cnlangzi commented on zboxcore/sdk/commitworker.go:
```diff
@@ -226,8 +226,8 @@ func (req *CommitRequest) commitBlobber(rootRef *fileref.Ref, latestWM *marker.W
 	}
-	formWriter.WriteField("connection_id", req.connectionID)
-	formWriter.WriteField("write_marker", string(wmData))
+	_ = formWriter.WriteField("connection_id", req.connectionID)
```
please use //nolint, or ignore it.

cnlangzi commented on zboxcore/sdk/copyworker.go:
```diff
@@ -45,8 +46,8 @@ func (req *CopyRequest) copyBlobberObject(blobber *blockchain.StorageNode, blobb
 	_ = formWriter.WriteField("connection_id", req.connectionID)
-	formWriter.WriteField("path", req.remotefilepath)
-	formWriter.WriteField("dest", req.destPath)
+	_ = formWriter.WriteField("path", req.remotefilepath)
```
please use //nolint or ignore it.

cnlangzi commented on core/zcncrypto/ed25519_test.go:
```diff
@@ -32,18 +33,20 @@ func TestEd25519GenerateKeys(t *testing.T) {
 	}
 	// Sign with valid private key
-	signScheme.SetPrivateKey(edsignPrivatekey)
+	_ = signScheme.SetPrivateKey(edsignPrivatekey)
```
please use //nolint or ignore it.

cnlangzi commented on zboxcore/sdk/sdk.go:
```diff
@@ -44,7 +44,7 @@ type StatusCallback interface {
 
 var numBlockDownloads = 10
-var sdkInitialized = false
+var SdkInitialized = false
```
is it unnecessary to export.

cnlangzi commented on zboxcore/sdk/sync.go:
```diff
@@ -189,6 +189,8 @@ func findDelta(rMap map[string]fileInfo, lMap map[string]fileInfo, prevMap map[s
 	rDelMap := make(map[string]string)
+	//S1005: unnecessary assignment to the blank identifier
+	//nolint
 	for rPath, _ := range rMap {
```
is it `for rPath := range rMap`?

Rizary replied:
yes.

cnlangzi commented on zcncore/wallet.go:
```diff
@@ -244,7 +244,7 @@ func GetMinShardersVerify() int {
 func getMinShardersVerify() int {
 	minSharders := util.MaxInt(calculateMinRequired(float64(_config.chain.MinConfirmation), float64(len(_config.chain.Sharders))/100), 1)
-	Logger.Info("Minimum sharders used for verify :", minSharders)
+	// Logger.Info("Minimum sharders used for verify :", minSharders)
```
why it is removed?

Rizary replied:
if we enabled this, the wallet_test.go get stuck, so when debugging the test code, I remove logger and test worked again. Other Logger.Info is find, I'm not sure why it get stuck with this function.

cnlangzi replied:
it is so strange. It should have a corner issue. Logger.Info should be not the bug. Could you try to test it more, and find what happens?

moldis replied:
@Rizary I guess bcz you need to init Logger in tests, but we should not change main code

Rizary replied:
The problem is not that apparently. It seems that `_config.chain.MinConfirmation` is still `nil` even though I already run `InitStorageSDK` which also run `zcncore.Init()`. In general, it should get updated. I'm still looking for solution.

Rizary commented on Makefile:
```diff
@@ -26,13 +26,16 @@ clean-gosdk:
+wasm-test:
+	CGO_ENABLED=0 GOOS=js GOARCH=wasm $(GOROOT)/bin/go test -v github.com/0chain/gosdk/wasm
 
-gosdk-test:
```
I'm not replacing with wasm-test, I have `gosdk-test: wasm-test` in the below

cnlangzi replied:
by the way, you don't need $(GOROOT)/bin/go , it should be ready on $PATH. so just run it via `go`

cnlangzi replied:
if it doesn't work on your local. please fix path issue. https://golang.org/doc/install/source#install

Rizary commented on Makefile:
```diff
@@ -57,8 +60,11 @@ test: gosdk-test
 clean: clean-gosdk clean-herumi
 	@rm -rf $(OUTDIR)
 
-lint:
```
I have add `lint: lint-wasm` so no, I dont remove anything.

Rizary commented on Makefile:
```diff
@@ -26,13 +26,16 @@ clean-gosdk:
 gosdk-build: gomod-download
 	go build -x -v -tags bn256 ./...
 
-sdkver:
```
`sdkver` is now `wasm` package, so everything now inside `wasm` folder which is self-explanatory.

Rizary commented on zboxcore/fileref/fileref.go:
```diff
@@ -57,7 +57,7 @@ func (a *Attributes) Validate() (err error) {
 }
 
 type FileRef struct {
-	Ref                 `json:",squash"`
```
I got the lint error saying `squash` is not json option.

cnlangzi replied:
it is added for `https://github.com/mitchellh/mapstructure`.  so please don't change anything if you are not sure why it is added there.

cnlangzi replied:
sometimes golang-lint is stupid

Rizary commented on zboxcore/encoder/erasurecode.go:
```diff
@@ -10,10 +10,10 @@ import (
-	Encode(in []byte) ([][]byte, error)
-	Decode(in [][]byte) ([]byte, error)
-}
+// type codec interface {
```
noted.

Rizary commented on wasm/wallet.go:
```diff
@@ -224,10 +217,12 @@ func IsMnemonicValid(this js.Value, p []js.Value) interface{} {
 
 func SetWalletInfo(this js.Value, p []js.Value) interface{} {
```
no, we can't do that. Please refer to `https://pkg.go.dev/syscall/js#FuncOf`

cnlangzi replied:
ok

Rizary commented on wasm/wallet.go:
```diff
@@ -1127,22 +1128,26 @@ func GetWritePoolInfo(this js.Value, p []js.Value) interface{} {
 func Encrypt(this js.Value, p []js.Value) interface{} {
 	key := p[0].String()
-	text := p[0].String()
+	text := p[1].String()
```
see above

Rizary commented on wasm/wallet.go:
```diff
@@ -1127,22 +1128,26 @@ func GetWritePoolInfo(this js.Value, p []js.Value) interface{} {
 func Decrypt(this js.Value, p []js.Value) interface{} {
 	key := p[0].String()
-	text := p[0].String()
+	text := p[1].String()
```
see above.

Rizary commented on zboxcore/client/entity.go:
```diff
@@ -7,8 +7,8 @@ import (
 type Client struct {
-	*zcncrypto.Wallet
-	signatureSchemeString string
+	Wallet                *zcncrypto.Wallet `json:"wallet"`
```
because I think we need to have `{wallet: { }, signature_scheme: {}}` for clarity? or go already did that for us?

cnlangzi replied:
so you don't export them. I have got it fixed on other pr. just update init with
```
func init(){
 client := &client{
     wallet:&zcncrype.Wallet{}
})
  
```

cnlangzi replied:
We can't change a lot of code to let test passed without any fix. Because it works on testnet. It should be fixed by `setup`

Rizary commented on zboxcore/client/entity.go:
```diff
@@ -7,8 +7,8 @@ import (
-	*zcncrypto.Wallet
-	signatureSchemeString string
+	Wallet                *zcncrypto.Wallet `json:"wallet"`
+	SignatureSchemeString string            `json:"signature_scheme"`
```
see above.

Rizary commented on zboxcore/sdk/allocation.go:
```diff
@@ -222,15 +222,15 @@ func (a *Allocation) InitAllocation() {
 
 func (a *Allocation) isInitialized() bool {
-	return a.initialized && sdkInitialized
+	return a.initialized && SdkInitialized
```
when we run `InitStorageSDK`, normally `sdkInitialized` is set to true. However, in our WASM version, I can't pass this test (cfm. https://github.com/0chain/gosdk/blob/rizary-wallet-zboxsdk/wasm/sdk_test.go#L160) because the `InitStorageSDK` doesn't set the `sdkInitialized` to true. Whereas, the `blockchain` is changed, and other test pass, the `TestReadPool` fail because it's think that sdk is not initialized.

In our other test, like `InitZCNSDK`, it works well.

cnlangzi replied:
you need more `setup` action on your test. so please don't change anything without any bugfixs , just get your test passed. It may break somethings

Rizary commented on zboxcore/sdk/allocation.go:
```diff
@@ -466,7 +466,7 @@ func (a *Allocation) downloadFile(localPath string, remotePath string, contentMo
 	}
 	lPath, _ := filepath.Split(localPath)
-	os.MkdirAll(lPath, os.ModePerm)
+	_ = os.MkdirAll(lPath, os.ModePerm)
```
it's pass with //nolint.

shaktals commented on bls/bls.go:
```diff
@@ -215,7 +216,7 @@ func (sig *Sign) Recover(shares []Sign, from []ID) error {
 		_a.Div(b)
 		t := BN254.G1mul(shares[i].v, _a) // G::mul(shares[i], a/b)
-		r.Add(t) // G::add(r, t) // r += t
+		r.Add(t)                          // G::add(r, t) // r += t
```
This seems not necessary?

moldis commented on core/zcncrypto/bls0chain_test.go:
````diff
@@ -115,20 +115,21 @@ func TestSerialization(t *testing.T) {
 // sk.SetByCSPRNG()
 // pk := sk.SerializeToHexStr()
 // ```
+//nolint
````
@cnlangzi Why do we need this directive? Just exclude tests files from lint check, otherwise, code becomes messy with` //nolint` directive which is become useless for me

moldis commented on wasm/minver/proxy_min.go:
```diff
@@ -912,70 +913,77 @@ func main() {
-	js.Global().Set("GetWritePoolInfo", js.FuncOf(GetWritePoolInfo))
-	js.Global().Set("Encrypt", js.FuncOf(Encrypt))
-	js.Global().Set("Decrypt", js.FuncOf(Decrypt))
+	// js.Global().Set("GetMinShardersVerify", js.FuncOf(GetMinShardersVerify))
```
Pls clean commented code if it's not used

moldis commented on zboxcore/sdk/sdk.go:
```diff
@@ -83,7 +83,7 @@ func InitStorageSDK(clientJson string, blockWorker string, chainID string, signa
 
 	go UpdateNetworkDetailsWorker(context.Background())
-	sdkInitialized = true
+	SdkInitialized = true
```
@Rizary We should not change main sdk code in this branch

moldis commented on wasm/sdk.go:
```diff
@@ -0,0 +1,800 @@
+	return nil
+}
+
+// //
```
@Rizary Pls cleanup all comments, they should be only on Top of the function explaining what this function is for.

moldis commented on wasm/sdk.go:
```diff
@@ -0,0 +1,800 @@
+			responseConstructor := js.Global().Get("Response")
+			response := responseConstructor.New(js.ValueOf(err))
+
+			// Resolve the Promise
```
Pls, remove these meaningless comments in functions body

moldis commented on wasm/sdk.go:
```diff
@@ -0,0 +1,800 @@
+		return nil
+	})
+
+	// Create and return the Promise object
```
+ 1 comments

moldis commented on wasm/sdk.go:
```diff
@@ -0,0 +1,800 @@
+// // for current client of the sdk.
+func ZBOXGetStakePoolInfo(this js.Value, p []js.Value) interface{} {
+	blobberID := p[0].String()
+	handler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
```
I can see lots of duplicated code especially with `js.FuncOf` . Pls find a way to avoid it

Rizary replied:
Ok noted, I will investigate it further.

moldis commented on wasm/sdk.go:
```diff
@@ -0,0 +1,800 @@
+		reject := args[1]
+
+		go func() {
+			err := initSDK(clientJSON, chainJSON)
```
That will cause an issue, as we Init SDK outside of this function

Rizary replied:
Should we just remove all of he `initSDK` inside the SDK function? Assuming we should run `initSDK` before running all of the function? IF yes, then I should also remove the `initSDK` from the `zboxcore/sdk/sdk.go` .

moldis commented on wasm/sdk.go:
```diff
@@ -0,0 +1,800 @@
+}
+
+func GetAllocationFromAuthTicket(this js.Value, p []js.Value) interface{} {
+	authTicket := p[0].String()
```
I think this will not work in `wasm` library

moldis commented on wasm/sdk.go:
```diff
@@ -0,0 +1,800 @@
+
+func ZBOXGetAllocation(this js.Value, p []js.Value) interface{} {
+	allocationID := p[0].String()
+	result, err := sdk.GetAllocation(allocationID)
```
I think this will not work in `wasm` library

moldis commented on wasm/sdk.go:
```diff
@@ -0,0 +1,800 @@
+}
+
+func ZBOXGetAllocations(this js.Value, p []js.Value) interface{} {
+	result, err := sdk.GetAllocations()
```
I think this will not work in `wasm` library

moldis commented on wasm/sdk.go:
```diff
@@ -0,0 +1,800 @@
+	readPrice := strToPriceRange(s_read)
+	writePrice := strToPriceRange(s_write)
+
+	result, err := sdk.CreateAllocation(datashards, parityshards, size, expiry, readPrice, writePrice, mcct, lock)
```
I think this will not work in `wasm` library

moldis commented on wasm/sdk.go:
```diff
@@ -0,0 +1,800 @@
+	readPrice := strToPriceRange(s_read)
+	writePrice := strToPriceRange(s_write)
+
+	result, err := sdk.CreateAllocationForOwner(owner, ownerpublickey, datashards, parityshards, size, expiry, readPrice, writePrice, mcct, lock, preferredBlobbers)
```
I think this will not work in `wasm` library

moldis commented on wasm/sdk.go:
```diff
@@ -0,0 +1,800 @@
+	allocationID := p[2].String()
+	lock, _ := strconv.ParseInt(p[3].String(), 10, 64)
+
+	result, err := sdk.UpdateAllocation(size, expiry, allocationID, lock)
```
I think this will not work in `wasm` library

moldis commented on wasm/sdk.go:
```diff
@@ -0,0 +1,800 @@
+func FinalizeAllocation(this js.Value, p []js.Value) interface{} {
+	allocID := p[0].String()
+
+	result, err := sdk.FinalizeAllocation(allocID)
```
I think this will not work in `wasm` library

moldis commented on wasm/sdk.go:
```diff
@@ -0,0 +1,800 @@
+func CancelAllocation(this js.Value, p []js.Value) interface{} {
+	allocID := p[0].String()
+
+	result, err := sdk.CancelAllocation(allocID)
```
I think this will not work in `wasm` library

moldis commented on wasm/sdk.go:
```diff
@@ -0,0 +1,800 @@
+	s_blob := p[0].String()
+	blob := strToBlob(s_blob)
+
+	result, err := sdk.UpdateBlobberSettings(&blob)
```
I think this will not work in `wasm` library

moldis commented on wasm/sdk.go:
```diff
@@ -0,0 +1,800 @@
+	metaTxnData := p[0].String()
+	fabricConfigJSON := p[1].String()
+
+	result, err := sdk.CommitToFabric(metaTxnData, fabricConfigJSON)
```
I think this will not work in `wasm` library

moldis commented on wasm/sdk.go:
```diff
@@ -0,0 +1,800 @@
+	readPrice := strToPriceRange(s_read)
+	writePrice := strToPriceRange(s_write)
+
+	result, err := sdk.GetAllocationMinLock(datashards, parityshards, size, expiry, readPrice, writePrice, mcct)
```
I think this will not work in `wasm` library

moldis commented on wasm/sdk_test.go:
```diff
@@ -0,0 +1,174 @@
+var chainConfig = "{\"chain_id\":\"0afc093ffb509f059c55478bc1a60351cef7b4e9c008a53a6cc8241ca8617dfe\",\"block_worker\":\"http://127.0.0.1:1/dns\",\"miners\":[\"http://127.0.0.1:1/miner01\"],\"sharders\":[\"http://127.0.0.1:1/sharder01\"],\"signature_scheme\":\"bls0chain\",\"min_submit\":50,\"min_confirmation\":50,\"confirmation_chain_length\":3,\"eth_node\":\"https://ropsten.infura.io/v3/f0a254d8 ...
+
+func TestInitializeConfig(t *testing.T) {
+	setup(t)
```
For testing please use sub-tests, like t.Run("test name", t {}) ... it will avoid many duplications and confusions

moldis commented on wasm/wallet_test.go:
```diff
@@ -0,0 +1,209 @@
+
+func TestNetwork(t *testing.T) {
+	setup(t)
+	defer server.Close()
```
Same issue for subtests

moldis commented on wasm/wallet_test.go:
```diff
@@ -0,0 +1,209 @@
+	getNetworkJSON := js.FuncOf(GetNetworkJSON)
+	defer getNetworkJSON.Release()
+
+	if got := getNetworkJSON.Invoke().String(); got != fmt.Sprintf("{\"miners\":[%#v],\"sharders\":[%#v]}", miner.URL+"/miner01", sharder.URL+"/sharder01") {
```
tests suppose to have asserts, but not error responses

moldis commented on wasm/wallet_test.go:
```diff
@@ -0,0 +1,209 @@
+	wi := setWalletInfo.Invoke(walletConfig, "true")
+
+	if got := wi.IsNull(); !got {
+		t.Errorf("Error: %#v", wi.Get("error").String())
```
+1 asserts

moldis reviewed (changes requested):
Need changes

shaktals commented:
@Rizary one important question, what is the final size of the `wasm` binary?

Rizary commented:
@shaktals it's around 14MB. I haven't created it again since right now I'm focusing on fixing the test. I'll let you know once I fix all the required changes.

shaktals commented:
> @shaktals it's around 14MB. I haven't created it again since right now I'm focusing on fixing the test. I'll let you know once I fix all the required changes.

@Rizary that is a large file to download when using a webapp, specially if you are on a _not so fast_ mobile network, which  unfortunately is still very common :sweat_smile:. Can we create a trimmed down `wasm` module only with the essential logic for the `js-client-sdk`/`block-explorer`?

Rizary commented:
@moldis My current commit finished:
1.  changed all the HTTP request function to use Promise based handler in WASM
2. Adding subtest
3. fixing some return error message
4. Debugging some test failure

My current blocker right now:
1. Some value is not set when initiating StorageSDK and it causes failure in the test. I'm still investigating the cause
2. When running `SetWalletInfo`, the wallet info is not updated, so some test fail unless I run `SetWalletInfo` inside that test block.

Rizary commented:
@cnlangzi I've test the Makefile, it works fine. I am still leaving some zcncore function that has similar name with zboxcore/sdk until all tests are passed.

Rizary commented:
> > @shaktals it's around 14MB. I haven't created it again since right now I'm focusing on fixing the test. I'll let you know once I fix all the required changes.
> 
> @Rizary that is a large file to download when using a webapp, specially if you are on a _not so fast_ mobile network, which unfortunately is still very common . Can we create a trimmed down `wasm` module only with the essential logic for the `js-client-sdk`/`block-explorer`?

Sure, there is a `minver` package inside the wasm folder to create small version of wasm. I'll cherry pick the function and let you know accordingly.

moldis commented on zboxcore/sdk/downloadworker.go:
```diff
@@ -93,7 +93,7 @@ func (req *DownloadRequest) downloadBlock(blockNum int64, blockChunksMax int) ([
 		encscheme = encryption.NewEncryptionScheme()
 		// TODO: Remove after testing
-		encscheme.Initialize(client.GetClient().Mnemonic)
+		encscheme.Initialize(client.GetClient().Wallet.Mnemonic)
```
@Rizary what is the reason of changing this?

Rizary replied:
I'll remove it in the next commit. It is because I want to make the Wallet public. But it turns out already being fixed in the `init()` function.

moldis commented on zboxcore/client/entity.go:
```diff
@@ -8,13 +8,15 @@ import (
 
 func init() {
-	client = &Client{}
+	client = &Client{
```
@Rizary What's the reason of changing this?

Rizary replied:
@cnlangzi has it fixed in master. Basically because it will caused null pointer when we call wallet.

moldis commented on wasm/wallet_test.go:
```diff
@@ -0,0 +1,178 @@
+	Logger.Info("Testing WASM version of Zcncore.Wallet")
+	defer server.Close()
+	t.Run("Setting All Configuration", func(t *testing.T) {
+		TestAllConfig(t)
```
The idea of subtests is not to call another function, but to reduce the amount of functions. Test code should be inside subtest

Rizary replied:
I fixed it.

moldis commented on wasm/sdk_test.go:
```diff
@@ -0,0 +1,88 @@
+
+		setMaxTxnQuery.Invoke("1")
+
+		assert.Equal(t, 1, blockchain.GetMaxTxnQuery())
```
You have to assert response from function as well

Rizary replied:
I fixed it in the latest commit.

moldis commented on wasm/sdk.go:
```diff
@@ -0,0 +1,1178 @@
+	return result
+}
+
+func SetSDKLogLevel(this js.Value, p []js.Value) interface{} {
```
Don't need this function in wasm

moldis commented on wasm/sdk.go:
```diff
@@ -0,0 +1,1178 @@
+	return nil
+}
+
+func SetSDKLogFile(this js.Value, p []js.Value) interface{} {
```
Don't need this function in wasm

moldis commented on wasm/sdk.go:
```diff
@@ -0,0 +1,1178 @@
+}
+
+// //
+// // read pool
```
Comments still not cleaned

moldis commented on wasm/sdk.go:
```diff
@@ -0,0 +1,1178 @@
+	return nil
+}
+
+func ZBOXGetReadPoolInfo(this js.Value, p []js.Value) interface{} {
```
Please, refactor names without ZBOX

Rizary replied:
@moldis It has name collition with the `wasm/wallet.go` function. What do you suggest? Because the function should be the same as in the `zboxcore/sdk.go` function which is `GetReadPoolInfo` .

Rizary replied:
So I removed all `ZBOX` word in `wasm/sdk.go` function and changed all function collision in `wasm/wallet.go` by adding `Wallet` word after the Get/Set .

moldis commented on wasm/sdk.go:
```diff
@@ -0,0 +1,1178 @@
+}
+
+// //
+// // stake pool
```
+1 Comments cleanup

moldis commented on wasm/sdk.go:
```diff
@@ -0,0 +1,1178 @@
+	return promiseConstructor.New(handler)
+}
+
+// // GetStakePoolUserInfo obtains blobbers/validators delegate pools statistic
```
Function header comments format is: // GetStakePoolUserInfo but not // // GetStakePoolUserInfo . Please refactore everywhere

Rizary replied:
I remove all comment in the `wasm/sdk.go` because it is already in the `zboxcore/sdk/sdk.go`

moldis commented on wasm/sdk.go:
```diff
@@ -0,0 +1,1178 @@
+	return promiseConstructor.New(handler)
+}
+
+// //
```
Comments cleanup

moldis commented on wasm/minver/zboxsdk_min.go:
```diff
@@ -0,0 +1,90 @@
+}
+
+func ZBOXGetClientEncryptedPublicKey(this js.Value, p []js.Value) interface{} {
+	result, err := sdk.GetClientEncryptedPublicKey()
```
@Rizary Have you tested this calls? I am sure they will not work as they access http calls.

moldis replied:
Same for functions below

Rizary replied:
I will update the min version which already updated in the wasm package.

moldis reviewed (dismissed):
Need refactoring

Rizary commented:
@moldis Thank you for taking the time to review this PR. Based on my latest commit, I have fixed all your requirement (please read my replies in your review). Thus, all the test that are error in the previous commit is now PASS. 

The `GetMinSharedVerify` function is no longer stuck (I fixed it by changing the function into Promise style function). All error behavior from the previous commit is fixed.

@shaktals I already cleaned up the minimun version of wasm so its output now down from 14MB to just 1MB. It still needs to be tested now. Please let me know if you got any error in the function.

In order to build the minimum version, please run the following command:

```bash
CGO_ENABLED=0 GOOS=js GOARCH=wasm go build -tags minver -o min.wasm github.com/0chain/gosdk/wasm
```

moldis commented on wasm/all_test.go:
```diff
@@ -0,0 +1,94 @@
+package wasm
```
@Rizary name if the test file is incorrect. Tests should be attached to a specific file

Rizary replied:
Should it? I mean we can have one test that might be imported in another test files, right? Because it's kinda redundant if I write setup twice. Or it can be confusing if we put setup in one of the test file. What do you suggest?

moldis replied:
_test.go should belong to something, not just an abstract test. You can move these tests to one of the proxy_test or sdk_test if they not belong to any actuall file

moldis reviewed (approved):
Looks fine, just a minor comment to resolve and then can merge to jssdk. After that please create PR jssdk -> staging to see if something missed

Rizary commented on wasm/minver.go:
```diff
@@ -0,0 +1,50 @@
+	js.Global().Set("SetNumBlockDownloads", js.FuncOf(SetNumBlockDownloads))
+	js.Global().Set("GetAllocations", js.FuncOf(GetAllocations))
+	js.Global().Set("GetAllocationFromAuthTicket", js.FuncOf(GetAllocationFromAuthTicket))
+	<-c
```
yes.

shaktals commented:
> @shaktals I already cleaned up the minimun version of wasm so its output now down from 14MB to just 1MB. It still needs to be tested now. Please let me know if you got any error in the function. The minimum version is now lived in the `wasm/minver.go`
> 
> In order to build the minimum version, please run the following command:
> 
> ```shell
> CGO_ENABLED=0 GOOS=js GOARCH=wasm go build -tags minver -o min.wasm github.com/0chain/gosdk/wasm
> ```

@Rizary It'd be nice to have this documented on README, or somewhere.

cnlangzi reviewed (approved):
LGTM.
